util/flags.py

```diff
@@ -100,7 +99,14 @@ def create_flags():
     f.DEFINE_string('checkpoint_dir', '', 'directory in which checkpoints are stored - defaults to directory "deepspeech/checkpoints" within user\'s data home specified by the XDG Base Directory Specification')
     f.DEFINE_integer('checkpoint_secs', 600, 'checkpoint saving interval in seconds')
     f.DEFINE_integer('max_to_keep', 5, 'number of checkpoint files to keep - default value is 5')
-    f.DEFINE_string('load', 'auto', '"last" for loading most recent epoch checkpoint, "best" for loading best validated checkpoint, "init" for initializing a fresh model, "auto" for trying the other options in order last > best > init')
+    f.DEFINE_string('load', 'auto', '"last" for loading most recent epoch checkpoint, "best" for loading best validated checkpoint, "init" for initializing a fresh model, "transfer" for transfer learning, "auto" for trying the other options in order last > best > init')
```

Comment:
I understand we want to limit the amount of flags, and I'm wondering if it's a good idea to have the `load` triggering different behaviors. Could we derive that knowledge of `transfer` from another flag ?

Reply:
we could trigger from `--source_model_checkpoint_dir`?

Reply:
I guess we could, but I have not read everything yet, so I'm unsure if that might break some of your usecases ?

Reply:
I've added `--save_checkpoint_dir` and `--load_checkpoint_dir`

Reply:
Are we fine breaking the compatibility here? @reuben Each and every user of DeepSpeech will get tricked. And it means for non-learning-transfer use-cases we will have to set **two** values, which might feel un-natural and I'm pretty sure this will translate in issues and discourse rant

Reply:
@lissyx we have too many checkpoint loading cases:

|From|To|Transfer|
|---|---|---|
|CuDNN|CuDNN|No|
|CuDNN|CuDNN|Yes|
|CuDNN|LSTMBlockFusedCell|No|
|CuDNN|LSTMBlockFusedCell|Yes|
|LSTMBlockFusedCell|LSTMBlockFusedCell|No|
|LSTMBlockFusedCell|LSTMBlockFusedCell|Yes|
|LSTMBlockFusedCell|CuDNN|No|
|LSTMBlockFusedCell|CuDNN|Yes|

I don't even know if some of these cases work before or after this PR, as I've never tried them (loading a LSTMBlockFusedCell checkpoint onto a CuDNN graph).

The way I see it, the motivation for the separate `load` and `save` flags is that some of these operations are seen as "more lossy" than others, or at least more dangerous. There are three general operations:

1. Continuing a run
2. Transfer learning from a checkpoint
3. Convert a CuDNN checkpoint to LSTMBlockFusedCell (or vice-versa)

I'd say most users have different expectations of data loss when doing operation 1 compared to 2 and 3. I transfer/convert something from X to Y, there's an expectation of a source and a destination even in the language used to describe these operations.

Users are already hitting problems here like mixing up CuDNN and LSTMBlockFusedCell, so I'd say it's pretty much guaranteed the transfer learning is going to cause people to get confused. But I don't think keeping it to one flag is going to avoid that. I think the benefit of being lossless in all causes outweighs the added complexity of having to specify separate load and save dirs for common cases.

Reply:
Maybe the least pitfally UI for this would be a separate tool that *only* converts a checkpoint. Example:

```
$ python convert_checkpoint.py --from_dir /a --from_type cudnn --to_dir /b --to_type lstmblockfusedcell --drop_source_layers 2
$ python train.py --checkpoint_dir /b --nouse_cudnn --n_hidden 2048 ...
```

But this requires quite a bit of work to make the training stuff more easily usable from outside.

Reply:
@reuben's proposal for a checkpoint converter is on-point.

`Deepspeech.py` has become too big, IMO

however, this PR is turning into a massive overhaul of checkpointing, but it is supposed to be transfer-learning specific.

before we make a `convert_checkpoint.py` script, I propose in this PR we add back the `--checkpoint_dir` flag, and when it is used, set `--load_checkpoint_dir` equal to `--save_checkpoint_dir`. The latter two flags will only be used by people who know how to use them

Reply:
Agreed on all points. And the backwards compatible `--checkpoint_dir` sounds like a great way to make this transition more seamless.

Reply:
I agree with all your comments.